eventlet/dagpool.py

```diff
@@ -0,0 +1,484 @@
+        available for any greenthread that depends on any of those keys.
+        """
+        try:
+            # If a dict is passed, copy it. Don't risk a subsequent
+            # modification to passed dict affecting our internal state.
+            iteritems = preload.iteritems()
+        except AttributeError:
+            # Not a dict, just an iterable of (key, value) pairs
+            iteritems = preload
+
+        # Load the initial dict
+        self.values = dict(iteritems)
+
+        # track greenthreads
+        self.coros = {}
+
+        # The key to blocking greenthreads is the Event.
+        self.event = Event()
+
+    def wait_each(self, keys):
+        """
+        keys is an iterable of keys.
+
+        wait_each(keys) is a generator producing (key, value) pairs as a value
+        becomes available for each requested key. wait_each() blocks the
+        calling greenthread until values become available. If the DAGPool was
+        prepopulated with any values, of course those can be delivered
+        immediately without waiting.
+
+        As soon as each key acquires a value, wait_each() yields a (key,
+        value) pair. Delivery order is intentionally decoupled from the
+        initial sequence of keys: each value is delivered as soon as it
+        becomes available. If multiple keys are available at the same time,
+        wait_each() delivers each of the ready ones in arbitrary order before
+        blocking again.
+
+        The DAGPool does not distinguish between a value returned by one of
+        its own greenthreads and one provided by a post() call.
+
+        The wait_each() generator terminates (raises StopIteration) when all
+        specified keys have been delivered. Thus, typical usage might be:
+
+        for key, value in dagpool.wait_each(keys):
+            # process this ready key and value
+        # continue processing now that we've gotten values for all keys
+
+        By implication, if you pass wait_each() an empty iterable of keys, it
+        returns immediately without yielding anything.
+        """
+        # Build a local set() and then call _wait_each().
+        return self._wait_each(set(keys))
+
+    def _wait_each(self, pending):
+        """
+        pending is a set() of keys for which we intend to wait. THIS SET WILL
+        BE DESTRUCTIVELY MODIFIED: as each key acquires a value, that key will
+        be removed from the passed 'pending' set.
+
+        In all other respects, _wait_each() behaves as described for wait_each().
+        """
+        while True:
+            # Before even waiting, show caller any (key, value) pairs that
+            # are already available. Copy 'pending' because we want to be able
+            # to remove items from the original set while iterating.
+            for key in pending.copy():
+                try:
+                    value = self.values[key]
+                except KeyError:
+                    pass
```

Comment:
It's your call, but I really think that easier to read (also faster but it's a joke in Python)

```
# somewhere in top
_missing = object()


value = self.values.get(key, _missing)
if value is not _missing:
  ...
```

Same applies to all following blocks of code that catch KeyError.